presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregatesIntoTableScanRuleSet.java

```diff
@@ -0,0 +1,190 @@
+                    .collect(toList());
+
+            Optional<AggregatePushdownResult<TableHandle>> connectorTableHandleAggregatePushdownResult = metadata.applyAggregation(context.getSession(),
+                    tableScan.getTable(),
+                    aggregateFunctions,
+                    assignments,
+                    Optional.of(ImmutableList.of(groupByColumns)));
+
+            if (!connectorTableHandleAggregatePushdownResult.isPresent()) {
+                return Result.empty();
+            }
+
+            AggregatePushdownResult<TableHandle> aggPushdownResult = connectorTableHandleAggregatePushdownResult.get();
+
+            TableHandle handle = aggPushdownResult.getHandle();
+            ImmutableList.Builder<Symbol> newScanOutputs = new ImmutableList.Builder<>();
+            ImmutableMap.Builder<Symbol, ColumnHandle> newScanAssignments = new ImmutableMap.Builder<>();
+            Map<String, ProjectionApplicationResult.Assignment> nameToAssignment = aggPushdownResult.getAssignments().stream()
+                    .collect(Collectors.toMap(ProjectionApplicationResult.Assignment::getVariable, Function.identity()));
+            for (Symbol aggOutputSymbol : node.getOutputSymbols()) {
+                ProjectionApplicationResult.Assignment assignment = nameToAssignment.get(aggOutputSymbol.getName());
+
+                if (assignment == null) {
+                    //TODO We pushdown all or nothing for V1 but we need to extend this in future.
+                    return Result.empty();
+                }
+                newScanOutputs.add(aggOutputSymbol);
+                newScanAssignments.put(aggOutputSymbol, assignment.getColumn());
+            }
+
+            return Result.ofPlanNode(
+                    TableScanNode.newInstance(
+                            tableScan.getId(),
+                            handle,
+                            newScanOutputs.build(),
+                            newScanAssignments.build()));
+        }
+    }
+
+    private ImmutableList<AggregateFunction> toAggFunction(Map<Symbol, AggregationNode.Aggregation> aggregations)
+    {
+        ImmutableList.Builder<AggregateFunction> aggregateFunctionBuilder = ImmutableList.builder();
+        for (Map.Entry<Symbol, AggregationNode.Aggregation> entry : aggregations.entrySet()) {
+            AggregationNode.Aggregation aggregation = entry.getValue();
+            Signature signature = aggregation.getResolvedFunction().getSignature();
+            Type returnType = typeManager.getType(signature.getReturnType());
+            String aggFunctionName = signature.getName();
+            List<Expression> arguments = aggregation.getArguments();
+            List<Symbol> symbols = arguments.stream()
+                    .filter(expression -> expression instanceof SymbolReference)
+                    .map(SymbolsExtractor::extractUnique)
+                    .flatMap(Collection::stream)
+                    .collect(toList());
```

Comment:
nit: `collect(toList())` -> `collect(toImmutableList())`

(here & in other places)

-- we prefer Guava immutable collections

Reply:
Done.